### ISSUE TITLE
Cloud Storage not persisting on error

### ISSUE THREAD

**Issue by techniq** [GoogleCloudPlatform/gcloud-node#1097]
I've ran into an issue that is only present in production using the Google Cloud Storage API.  I have an importer that loads 100 entries at a time.  When developing locally (with `require('fs')`) and an error occurs on say page 5, the previous pages worth of entries (400 total) are still persisted to disk.  When using GCS, this is not the case (the file doesn't even appear in the bucket).

Here is a snippet of the problematic part of the pipeline.
```js
    stream = importer.streamPages({perPage: perPage, importDate: importDate}, maxPages);
    
    let writeStream = null;
    if (process.env.NODE_ENV === 'production') {
      writeStream = bucket.file(importFileName).createWriteStream();
    } else {
      writeStream = fs.createWriteStream(`data/${importFileName}`);
    } 
    
    stream
      .on('error', err => this.logError(err))
      .pipe(ndjson.stringify())
      .pipe(writeStream);
```

**Comment by stephenplusplus** [comment]
The API can receive a file in two ways; a single request that contains the entire contents of the file (simple), or a series of requests (resumable). By default, `createWriteStream` uses resumable uploads, so that if an error occurs during the upload, you can start the transfer over, and only send data starting from where the last upload quit.

With a simple upload (you can pass `resumable: false` to `createWriteStream`), if an error occurs, the API never sees a completed HTTP request, so no file is ever created (*@jgeewax, is this right?*).

**Comment by techniq** [comment]
@stephenplusplus hmm, I'm using the default `resumable` stream.  If an error occurs, is there a way I can resume and just "commit" the current progress.

In my example, `importer.streamPages()` walks a paginated api, emitting a `perPage` amount of objects (it's an object stream) on each `data` event.  It would be nice if I could do something like this.

```js
stream
      .on('error', err => { this.logError(err); this.done(); )
```

**Comment by stephenplusplus** [comment]
That sounds like a cool idea. I'll see if I can find a way to do that :+1:

**Comment by stephenplusplus** [comment]
I think what you can do is:

```js
stream
      .on('error', err => { this.logError(err); this.end(); })
```

When a readable stream gracefully ends, it's saying "And you, next stream in the pipe, have received all I have to give you." Our write stream will continue normally, and write that data to the remote object. Try it out and let me know!

**Comment by techniq** [comment]
Thanks, I'll give that a shot and let you know.  In my example, I'll need to not sure an arrow function so `this` is the stream and not the closure, so I figure it will be more like...

```js
let logError = this.logError;
stream
      .on('error', function(err) { logError(err); this.end(); })
```

I currently don't have a test for this, but will try to make one (against Cloud Storage) and see how it goes, as otherwise I'll have to wait for the failure to occur again (`import.streamPages` retries a call 5 times before failing, which a gateway time-out caused the issue before)

**Comment by stephenplusplus** [comment]
Cool, thanks for sharing details about your use case. It always helps to hear real-world scenarios.

RE: the `this` stuff, good catch, I didn't think of that. Here's another option:

```js
stream
      .on('error', err => { this.logError(err); stream.end(); })
```

And just for reference, here is the script I used to test:

```js
'use strict';

var through = require('through2')
var gcloud = require('gcloud')({ projectId: '...' })
var gcs = gcloud.storage()
var bucket = gcs.bucket('...')
var file = bucket.file('new-file-' + Date.now())

var srcStream = through.obj(function (chunk, enc, next) {
  next(null, JSON.stringify(chunk))
})
var writeStream = file.createWriteStream()

srcStream.on('error', function () {
  this.end()
})

srcStream.pipe(writeStream)

srcStream.write({ a: 'b' })
srcStream.write({ a: 'b' })
srcStream.write({ a: 'b' })
srcStream.destroy(new Error('error'))
```

**Comment by techniq** [comment]
@stephenplusplus thanks for your help.  I've made some progress on this, but my limited knowledge of Node streams has me hitting a wall.

I found some of my issue to be in my `retryStream` function, that uses a Passthrough stream/transform that I was previously just logging "Max attempts reached" when the `streamFunc` reached it's maximum number of retries.

Now, calling `.end()` on the PassThrough stream fixes the problem I was having with the file not showing up in GCS (after adding an arbitrary 1000ms setTimeout() to my test before calling done() so the node/mocha process didn't terminate before gcloude-node had time to write the file up to GCS.  I tried using `setImmediate()` but it didn't work). 

Anyways, the problem I have is I really need to bubble an error out of `retryStream` through `streamPage` and `streamPages` so I can add an 'error' event listener at the end of the pipeline so I can properly log/email the problem before writing what I have grabbed so far to GCS (and a importStream which also processes the records and persists them to Cloud SQL) .

If I remove the error handler from `BaseImporter#streamPage`, then I get `Uncaught Error: Uncaught, unspecified "error" event. (Max attempts reached)`, but if I leave it on, the error gets swallowed and the test times out because `end` or `error` never makes it.

If I attempt to "re-emit" the error (`.on('error', function(err) { console.error('streamPage error', err); this.emit('error', err) })`) I get a ` Uncaught RangeError: Maximum call stack size exceeded`

This is quite likely out of the scope of this issue, but thought I'd at least present my problem in case you had any suggestions.

Here are snippets of the related functions.   Thanks for any help you have on this.


`retry-stream.js`
```js
'use strict';
const PassThrough = require('stream').PassThrough;

module.exports = function retryStream(streamFunc, retries) {
  var resultStream = new PassThrough();
  let attempt = 1;
  retries = retries || 3;

  function attemptStream() {
    let stream = streamFunc()
      .on('error', (err) => {
        console.warn(`Attempt (${attempt} of ${retries}) failed`, err);
        if (attempt < retries) {
          console.log('Retrying');
          attempt++;
          attemptStream();
        } else {
          console.error('Max attempts reached');
          // resultStream.end(); // <-- Works
          resultStream.emit('error', 'Max attempts reached'); // <-- Would like to work
        }
      })
      .on('response', () => {
        // console.log('success');
        stream.pipe(resultStream);
      })
  };
  attemptStream();

  return resultStream;
}
```

`base.js`
```js
const got = require('got');
const MultiStream = require('multistream');
const through2 = require('through2');
const JSONStream = require('JSONStream');
const JSONPath = require('jsonpath');
...

class BaseImporter {
...
  streamPage(query, options) {
    const pageOptions = this._getPageOptions(query, false);
    const queryString = Object.keys(pageOptions.query).map(key => `${key}=${pageOptions.query[key]}`).join('&');
    // console.log('Fetching url', `${this.config.baseUrl}?${queryString}`);
    
    if (options && options.retries) {
      return retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
        .on('error', err => console.error('streamPage error', err))
        .pipe(JSONStream.parse(this.config.jsonPath))
    } else {
      return got.stream(this.config.baseUrl, pageOptions)
        .on('error', err => console.error(err))
        .pipe(JSONStream.parse(this.config.jsonPath))
    }
  }

  streamPages(query, maxPages) {
    query.page = query.page || 1;
    let totalRaceCount = 0;
    let isDone = false;

    let factory = (streamCallback) => {
      if (isDone) {
        // console.log('Done');
        streamCallback(null);
        return;
      } else if (maxPages && query.page > maxPages) {
        console.log('Maximum pages reached.');
        streamCallback(null);
        return;
      } else {
        // console.log('Next page')
      }

      let raceCount = 0;
      let stream = this.streamPage(query, { retries: 5 })
        //.on('error', (err) => console.error('streamPages error', err))
        .on('end', () => {
          // console.error('streamPage end')
          console.log(`Fetched ${raceCount} races`);
          if (raceCount < query.perPage) {
            // console.log('Last page');
            isDone = true;
          }
          
          streamCallback(null, stream);
        })
        .pipe(through2({ objectMode: true, highWaterMark: query.perPage },
          function(chunk, enc, transformCallback) {
            // console.log('transform');
            // console.log('chunk', chunk);

            raceCount += 1;
            totalRaceCount += 1;

            this.push(chunk)
            transformCallback()
          })
        )

      query.page += 1;
    }

    return MultiStream(factory, { objectMode: true, highWaterMark: query.perPage })
      .on('error', (err) => console.error('MultiStream error', err))
      .on('close', () => console.log(`Total races: ${totalRaceCount}`))
  }
...
}
```

`test.js`
```js
describe.only('BaseImporter GCS', function() {

  beforeEach(function() {
    nock('http://localhost')
      .get('/')
        .query({page: 1, perPage: 5})
        .reply(200, (uri, requestBody) => {
          // console.log('returning page1, 200')
          return [{}, {}, {}, {}, {}]
        })
      .get('/')
        .query({page: 2, perPage: 5})
        .reply(200, (uri, requestBody) => {
          // console.log('returning page2, 200')
          return [{}, {}, {}, {}, {}]
        })
      .get('/')
        .query({page: 3, perPage: 5})
        .times(10)
        .reply(500, () => {
          // console.log('returning 500')
        })
  })

  afterEach(function() {
    nock.cleanAll();
  })

  describe('#streamPages()', function(done) {
    it('should still write contents to GCS for previous pages after exhausting retries for later page', function(done) {
      this.timeout(5000);
      
      const importer = new BaseImporter({ baseUrl: 'http://localhost', jsonPath: '*' });
      
      const gcloud = require('gcloud');
      const ndjson = require('ndjson');
      const gcsConfig = require('../../../app/config').get('storage:googleCloudStorage');
      const gcs = gcloud.storage({ projectId: gcsConfig.projectId });
      const bucket = gcs.bucket(gcsConfig.bucket);
      const importFileName = `test/${new Date().toISOString()}`;
      const writeStream = bucket.file(importFileName).createWriteStream();
      // const writeStream = require('fs').createWriteStream('test.txt');
      
      const stream = importer.streamPages({perPage: 5}, 3)
        .on(
        .on('end', () => {
          // Give GCS a second to upload file before ending test (and killing process)
          setTimeout(done, 1000);
        })
        .pipe(ndjson.stringify())
        .pipe(writeStream)

        ;
    });
  });

});
```

**Comment by stephenplusplus** [comment]
No problem, I'm happy to take a look! I'll let you know if I can come up with anything

**Comment by techniq** [comment]
@stephenplusplus you are crazy responsive :)

**Comment by stephenplusplus** [comment]
Also just repo-dropping this because I also go through some headaches when trying to deal with streams, so I wrote up some guides for under/non-documented use cases: https://github.com/stream-faqs - I haven't dug into your code yet, but it might have a helpful section or two!

**Comment by techniq** [comment]
You mean https://github.com/stephenplusplus/stream-faqs and I've been there - https://github.com/techniq/wiki/wiki/Node.js :)

**Comment by stephenplusplus** [comment]
Haha, I just make sure to always keep a GH tab open:

<img width="1440" alt="screen shot 2016-01-29 at 4 45 00 pm" src="https://cloud.githubusercontent.com/assets/723048/12688962/a9436ec2-c6a7-11e5-8915-51760ac3453e.png">

And oh, nice! :)

**Comment by techniq** [comment]
As well a nice game.  You have a daughter I image (or really like Frozen :) )

**Comment by techniq** [comment]
btw, I'm on Hangouts/Slack/etc if you want to chat outside of the issue to keep the noise down.  I'm leaving the office in about 10 min to work my second shift (Chief Family Officer) but will be back on ~9:00pm EST.

**Comment by stephenplusplus** [comment]
Sounds about identical to my plans! I'll ping back I'm back on the case. No
worries about noise here, I think it'll only help others.

And yes, a daughter who loves Frozen enough for both of us:
https://youtu.be/tGKybwbsA34

**Comment by stephenplusplus** [comment]
I'm not quite sure this is what you need, but `end` takes an error, first emitting it, then attempting to gracefully end the stream. So, instead of:

```js
// resultStream.end(); // <-- Works
resultStream.emit('error', 'Max attempts reached'); // <-- Would like to work
```

Try:

```js
resultStream.end(new Error('Max attempts reached'));
```

But for the issue of bubbling the error, I would try using [`pumpify`](https://github.com/mafintosh/pumpify).

I tried running the code, but I don't think I got it to work the way I was supposed to! I saw how removing the `setTimeout` caused the file not to be created, but I'm a little confused by the pipeline/process to really know where that problem is caused.

And just in case you haven't seen this one, take a look at [`retry-request`](https://github.com/stephenplusplus/retry-request)... it might save you the headache of the retry stuff.

And since all I'm doing is dropping module names, another: [`paged-http-stream`](https://github.com/karissa/paged-http-stream) might be nice here because of how much it simplifies the pagination process.

Sorry that this may have added more confusion than you were looking for :( If you can put together a gist or repo I can clone that gets me up and running a little easier, I'll be happy to dive back in.

**Comment by techniq** [comment]
Thanks @stephenplusplus, that gives me plenty to look at.  I've already had to switch gears onto a different project so it will probably be a few days till I can spend more time on this.  I've been using `got` (for Promise support, as well as stream) and originally it had retry support but not for 500 errors (which was odd to me), but I think they may have changed that in later releases to also retry on 500s.  If so, I might also try to look if it can be solved with it.

If I continue to hit a road block, I'll definitely try to get a working gist setup.  When I posted the snippets I tried to keep them concise to the issue but definitely omitted some parts that you probably need to get it functional.

Thanks again for all your help.

**Comment by stephenplusplus** [comment]
No problem. I'll close this issue since I don't think `gcloud` needs to change, your app just needs to find the right place to handle the aborting of the stream. I'm still happy to help find where that is, so whenever/whatever I can do, just ping me back in here and we can keep working on it.

**Comment by techniq** [comment]
@stephenplusplus No problem, and good news, I spent some time on this yesterday and figured it out.  My `retryStream` now emits a `new Error()` when it exhausts the number of retries.  One of the big gotchas (and still not 100% sure why it's occurring) was needing to call `stream.emit('error', err)` within my `streamPage` function instead of `this.emit('error', err)` even when using a normal function and not an arrow function.

If i used a function and `this.emit`, it would stack overflow (`Uncaught RangeError: Maximum call stack size exceeded`)
```js
      const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
        .on('error', function(err) { this.emit('error', err) })
        .pipe(JSONStream.parse(this.config.jsonPath))
```
but this works
```js
  const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
        .on('error', err => stream.emit('error', err))
        .pipe(JSONStream.parse(this.config.jsonPath))
```

I also refactored my `streamPages` to no longer use `MultiStream` and a factory function but instead rewrote it to just use a `PassThrough` stream as I do with `retryStream` (I didn't know `PassThrough` existed when I first wrote that and wasn't sure how to dynamically chain a `n` number of streams into 1).

Here are my current functions in case your interested.

`retry-stream.js`
```js
'use strict';
const PassThrough = require('stream').PassThrough;

module.exports = function retryStream(streamFunc, retries) {
  var resultStream = new PassThrough();
  let attempt = 1;
  retries = retries || 3;

  function attemptStream() {
    let stream = streamFunc()
      .on('error', (err) => {
        console.warn(`Attempt (${attempt} of ${retries}) failed`, err);
        if (attempt < retries) {
          // console.log('Retrying');
          attempt++;
          attemptStream();
        } else {
          // console.error('Max attempts reached');
          resultStream.emit('error', new Error('Max attempts reached'));
        }
      })
      .on('response', () => {
        // console.log('success');
        stream.pipe(resultStream);
      })
  };
  attemptStream();
  
  return resultStream;
}
```

```js
...
class BaseImporter {
...
  streamPage(query, options) {
    const pageOptions = this._getPageOptions(query, false);
    const queryString = Object.keys(pageOptions.query).map(key => `${key}=${pageOptions.query[key]}`).join('&');
    // console.log('Fetching url', `${this.config.baseUrl}?${queryString}`);
    
    if (options && options.retries) {
      const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
        // .on('error', function(err) { console.error('streamPage error', err); this.emit('error', err) })
        .on('error', err => stream.emit('error', err))
        .pipe(JSONStream.parse(this.config.jsonPath))
      return stream;
    } else {
      return got.stream(this.config.baseUrl, pageOptions)
        .on('error', err => console.error(err))
        .pipe(JSONStream.parse(this.config.jsonPath))
    }
  }

  streamPages(query, options) {
    const PassThrough = require('stream').PassThrough;
    let resultStream = new PassThrough({ objectMode: true, highWaterMark: query.perPage });
    
    let totalRaceCount = 0;

    const getPage = (page) => {
      let raceCount = 0;
      query.page = page;
      options = options || {};
      options.retries = options.retries || 5;
      let stream = this.streamPage(query, options)
        .on('error', (err) => {
          // console.error('streamPages error', err)
          resultStream.emit('error', err);
        })
        .on('end', () => {
          // console.log('streamPages end')
          
          console.log(`Fetched ${raceCount} races`);
          if (raceCount < query.perPage) {
            // console.log('Last page');
            resultStream.end();
          } else if (options.maxPages && page >= options.maxPages) {
            // console.log('Maximum pages reached.');
            resultStream.end();
          } else {
            // console.log('Next page')
            getPage(++page);
          }
        })
        .pipe(through2({ objectMode: true, highWaterMark: query.perPage },
          function(chunk, enc, transformCallback) {
            // console.log('transform');
            // console.log('chunk', chunk);

            raceCount += 1;
            totalRaceCount += 1;

            resultStream.write(chunk)
            transformCallback()
          })
        )
    };
    getPage(query.page || 1);
      
    return resultStream
      .on('error', err => console.log(`Error`, err, `total races at error: ${totalRaceCount}`))
      .on('end', () => console.log(`Total races: ${totalRaceCount}`))
  }
...
}
```

`base.test.js`
```js
describe('BaseImporter GCS', function() {

  describe('#streamPages()', function(done) {
    it('should still write contents to GCS for previous pages after exhausting retries for a later page', function(done) {
      this.timeout(5000);
      
      nock('http://localhost')
        .get('/')
          .query({page: 1, perPage: 5})
          .reply(200, (uri, requestBody) => {
            // console.log('returning page1, 200')
            return [{}, {}, {}, {}, {}]
          })
        .get('/')
          .query({page: 2, perPage: 5})
          .reply(200, (uri, requestBody) => {
            // console.log('returning page2, 200')
            return [{}, {}, {}, {}, {}]
          })
        .get('/')
          .query({page: 3, perPage: 5})
          .times(3)
          .reply(500, () => {
            // console.log('returning 500')
          })
      
      const importer = new BaseImporter({ baseUrl: 'http://localhost', jsonPath: '*' });
      
      const gcloud = require('gcloud');
      const ndjson = require('ndjson');
      const gcsConfig = require('../../../app/config').get('storage:googleCloudStorage');
      const gcs = gcloud.storage({ projectId: gcsConfig.projectId });
      const bucket = gcs.bucket(gcsConfig.bucket);
      const importFileName = `test/${new Date().toISOString()}`;
      const writeStream = bucket.file(importFileName).createWriteStream();
      // const writeStream = require('fs').createWriteStream('foo.txt');
      
      const stream = importer.streamPages({perPage: 5}, { retries: 3, maxPages: 5 })
        .on('error', err => {
          expect(err).to.exists
          expect(err.message).to.equal('Max attempts reached')
          stream.end()
          
          // Give GCS a second to upload file before ending test (and killing process)
          setTimeout(done, 1000);
        })
        .on('end', () => {
          expect.fail('"end" should not be emitted')
        })
        .pipe(ndjson.stringify())
        .pipe(writeStream)
        ;
    });
  });

});
```

**Comment by techniq** [comment]
and actually, I just tried to refactor...

```js
streamPage(query, options) {
    const pageOptions = this._getPageOptions(query, false);
    const queryString = Object.keys(pageOptions.query).map(key => `${key}=${pageOptions.query[key]}`).join('&');
    // console.log('Fetching url', `${this.config.baseUrl}?${queryString}`);
    
    if (options && options.retries) {
      const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
        .on('error', err => stream.emit('error', err))
        .pipe(JSONStream.parse(this.config.jsonPath))
      return stream;
    } else {
      return got.stream(this.config.baseUrl, pageOptions)
        .on('error', err => console.error(err))
        .pipe(JSONStream.parse(this.config.jsonPath))
    }
  }
```
to
```js
streamPage(query, options) {
    const pageOptions = this._getPageOptions(query, false);
    const queryString = Object.keys(pageOptions.query).map(key => `${key}=${pageOptions.query[key]}`).join('&');
    // console.log('Fetching url', `${this.config.baseUrl}?${queryString}`);
    
    let stream = (options && options.retries)
      ? retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
      : got.stream(this.config.baseUrl, pageOptions);
      
    return stream
      .on('error', err => stream.emit('error', err))
      .pipe(JSONStream.parse(this.config.jsonPath))
  }
```
and I'm back at getting the `Uncaught RangeError: Maximum call stack size exceeded`

**Comment by stephenplusplus** [comment]
Using this example:

```js
const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
    .on('error', function(err) { this.emit('error', err) })
    .pipe(JSONStream.parse(this.config.jsonPath))
```
but this works
```js
const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
    .on('error', err => stream.emit('error', err))
    .pipe(JSONStream.parse(this.config.jsonPath))
```

What's happening is in the first example, `this` is the same context that the error was emitted on. So it's going "I got an error? I better tell me about it! I got an error? I better tell me about it! I got an..."

In the second example, when you emit on `stream`, (this is the important part:) `stream` is the last stream in the pipe that was given when you assigned the variable `stream`. So `stream` is equal to `JSONStream.parse(this.config.jsonPath)`

Every time you call pipe, it's basically:

```js
Stream.prototype.pipe = function(nextStream) {
  // Glue this instance of a stream to `nextStream`...
  return nextStream;
};
```

So when you call `.pipe` after `.pipe`, the variable you're assigning them to will only reference the last one.

**Comment by techniq** [comment]
Interesting. So is there a better way to emit an error without it being catch by the same error handling?  I'm assuming this would fail (without testing it myself)

```js
const stream =  retryStream(() => got.stream(this.config.baseUrl, pageOptions), options.retries)
    .on('error', err => stream.emit('error', err))
```

**Comment by stephenplusplus** [comment]
If the error is a deal-breaker, you should end the stream from where the problem began by calling `this.end(err);` 

Here's an example of what happens when ending a source stream with an error:

```js
var through = require('through2')
var sourceStream = through()

stream
  .on('error', console.error.bind(console, '\nError on source stream'))
  .pipe(through())
  .on('error', console.error.bind(console, '\nError on destination stream'))
  .on('end', console.log.bind(console, 'ended'))
  .pipe(process.stdout)

sourceStream.push('a')
sourceStream.push('b')
sourceStream.push('c')
sourceStream.end(new Error(':('))
```

```sh
$ node .
abc
Error on source stream [TypeError: Invalid non-string/buffer chunk]
ended
```

In your case, the upload stream should see the peaceful end from your stream, and upload the data to GCS.

**Comment by techniq** [comment]
I had a problem when I tried to call `.end()` with an Error object instead of emitting an error event

Works
```js
...
resultStream.emit('error', new Error('Max attempts reached'));
...

expect(err.message).to.be('Max attempts reached')
```

Wrong err.message - `Invalid non-string/buffer chunk`
```js
...
resultStream.end(new Error('Max attempts reached'));

expect(err.message).to.be('Max attempts reached') // -> Uncaught AssertionError: expected 'Invalid non-string/buffer chunk' to equal 'Max attempts reached'
...
```

At this point I'm going to just go forward with what I have.  Node Streams are still perplexing with how they work to me, but at this point I have something I can work with.

Thanks again for all your help.

**Comment by stephenplusplus** [comment]
Oh man, you're totally right. I was mixing up `destroy` with `end`. I should just close up `stream-faqs` now and call it `stephen-is-a-noob`.

Alright, the example I should have given is:

```js
var through = require('through2')
var stream = through()

stream
  .on('error', console.error.bind(console, '\nError on source stream'))
  .pipe(through())
  .on('error', console.error.bind(console, '\nError on destination stream'))
  .on('end', console.log.bind(console, 'ended'))
  .pipe(process.stdout)

stream.push('a')
stream.push('b')
stream.push('c')
stream.emit('error', new Error(':('))
stream.end()
```

So emit the error like you are, but also call `end`, and the stream that you ended communicates to the other streams that they've received all the data they're going to get, so they can close properly. That should result in GCS getting the data.

Sorry to lead you down a dark path!

**Comment by techniq** [comment]
No worries.  Thanks :)

**Comment by techniq** [comment]
@stephenplusplus Guess who's back :)

We went into production with the changes we discussed here, but ran into a major issue.  
![screen shot 2016-02-02 at 11 32 12 am](https://cloud.githubusercontent.com/assets/177476/12756192/a2b89bb2-c9a0-11e5-9c47-c6c487450662.png)

The error handler is after the piping to the GCS write stream.

```js
...
    // Download and import new entries
    stream = importer.streamPages({perPage: perPage, importDate: importDate}, { retries: 5, maxPages: maxPages});

    let writeStream = null;
    if (process.env.NODE_ENV === 'production') {
      writeStream = bucket.file(importFileName).createWriteStream();
    } else {
      writeStream = fs.createWriteStream(`data/${importFileName}`);
    } 
    
    stream
      .pipe(ndjson.stringify())
      .pipe(writeStream)
      .on('error', err => {
        this.logError('Error writing stream', err, sourceName);
        stream.end();
      });
...
```

I've been unable to reproduce the issue locally on my workstation when using production resources (GCS, Google Cloud SQL, etc) via `NODE_ENV npm start`.  I think the following things might be highly related.

- We're currently using a `f1-micro` instance currently (so only 600MB mem)
- After the refactoring to `streamPages`, I don't think we're getting any back-pressure on the pipeline from the writeStream's (it's fetching 100 new races/records as fast as the event loop will allow it as far as I can tell from the logs).  I know this is due to how I'm performing the check within `.on('end', ...)` of the single `streamPage` call, but I'm not sure how to apply this logic so it will wait for the downstream writers/back-pressure before enqueuing the next `getPage/streamPage`.
- Hunch: We're blowing through the micro instances memory causing the socket hangup to GCS to occur

```js
  streamPages(query, options) {
    const PassThrough = require('stream').PassThrough;
    let resultStream = new PassThrough({ objectMode: true, highWaterMark: query.perPage });
    
    let totalRaceCount = 0;

    const getPage = (page) => {
      let raceCount = 0;
      query.page = page;
      options = options || {};
      options.retries = options.retries || 5;
      let stream = this.streamPage(query, options)
        .on('error', (err) => {
          // console.error('streamPages error', err)
          resultStream.emit('error', err);
          resultStream.end();
        })
        .on('end', () => {
          // console.log('streamPages end')
          
          console.log(`Fetched ${raceCount} races`);
          if (raceCount < query.perPage) {
            // console.log('Last page');
            resultStream.end();
          } else if (options.maxPages && page >= options.maxPages) {
            // console.log('Maximum pages reached.');
            resultStream.end();
          } else {
            // console.log('Next page')
            getPage(++page);
          }
        })
        .pipe(through2({ objectMode: true, highWaterMark: query.perPage },
          function(chunk, enc, transformCallback) {
            // console.log('transform');
            // console.log('chunk', chunk);

            raceCount += 1;
            totalRaceCount += 1;

            resultStream.write(chunk)
            transformCallback()
          })
        )
    };
    getPage(query.page || 1);
      
    return resultStream
      .on('error', err => console.log(`Error`, err, `total races at error: ${totalRaceCount}`))
      .on('end', () => console.log(`Total races: ${totalRaceCount}`))
  }
```

Any ideas?

**Comment by stephenplusplus** [comment]
Could you plop a gist or repo together with the bare minimums so I can jump in? (please include package.json!) I think we can figure it out, but I'm doing a lot of scrolling back and forth, so I think I'm probably just slowing us both down.

**Comment by techniq** [comment]
Will do.  Thanks

**Comment by techniq** [comment]
Probably take me an hour or so as I need to step into a meeter.  Are you on the Google Cloud team btw (just curious)?

**Comment by stephenplusplus** [comment]
Technically speaking, no. I'm only helping the team (@jgeewax) with this project.

**Comment by techniq** [comment]
I'm struggling to find a good way to reduce this into a simple reproducible example.  Since there are a lot of things at play (URL fetching of a large amount of data, slow consuming stream to a Cloud SQL database, and it's only reproducible (thus far) on a memory-limited micro instance.  You wouldn't be interested in doing a Hangout/screen share to walk through the code on my machine?

**Comment by stephenplusplus** [comment]
Sorry about the code back and forth. I was still trying to wrap my head around how all of the pieces of the pipeline played together. Stubs to stand-in for actual API services are fine. I was able to put the pieces together and get something running, but wasn't able to produce any meaningful results.

> I'm not sure how to apply this logic so it will wait for the downstream writers/back-pressure before enqueuing the next getPage/streamPage.

You shouldn't need to do this, as far as I know. The upload stream should only take on what it can handle.

My guess is it relates to the v1-micro instance, like you suggested. If not from the low memory, maybe another limitation on the amount of sockets allowed to remain open. Is it possible to try on a more equipped VM for sake of comparison? The default `maxSockets` allowed to be open should be Infinity, so it shouldn't be coming from the Node side (`require('http').Agent.defaultMaxSockets === Infinity`).

Sorry I can't be more help here as this goes deeper into the lower levels of the stack.

@jgeewax @callmehiphop @mziccard can you think of anything?

**Comment by techniq** [comment]
Good news, I went back to an implementation of `streamPages` that uses `MultiSteram` and I did not receive the `socket hang up` error during our nightly tasks.  Using MultiStream, streamPages's factory is only called after the downstream consumers (writing to GCS and importing into Cloud SQL) have completed the current batch.  I couldn't find any kind of memory usage graph on https://console.cloud.google.com to confirm my hunch, but I'm pretty sure continually downloading all the pages from the rest service was blowing through the f1-micro's small amount of memory.

I had moved away from MultiStream to try to simplify our stream pipeline when troubleshooting the originally cause for the issue (not writing to GCS if an error was raised from one of the API page calls).  With my new units tests in place, I was able to make the appropriate changes and get it all to work.  The big change was adding the following to the `streamPage()` (single page) call within the streams factory of MultiStream:

```js
        .on('error', function(err) {
          this.end();
          streamCallback(err);
        })
```

Here's the full impl if you're interested.  As always, thanks for the help.
```js
  streamPages(query, options) {
    query.page = query.page || 1;
    let totalRaceCount = 0;
    let isDone = false;

    let factory = (streamCallback) => {
      if (isDone) {
        console.log('Done');
        streamCallback(null);
        return;
      } else if (options.maxPages && query.page > options.maxPages) {
        console.log('Maximum pages reached.');
        streamCallback(null);
        return;
      }

      let raceCount = 0;
      let stream = this.streamPage(query, options)
        .on('error', function(err) {
          // console.error('streamPages error', err)
          this.end();
          streamCallback(err);
        })
        .on('end', () => {
          // console.error('streamPage end')
          console.log(`Fetched ${raceCount} races`);
          if (raceCount < query.perPage) {
            // console.log('Last page');
            isDone = true;
          } else {
            // console.log('Next page')
            query.page += 1;
          }
          
          streamCallback(null, stream);
        })
        .pipe(through2({ objectMode: true, highWaterMark: query.perPage },
          function(chunk, enc, transformCallback) {
            // console.log('transform');
            // console.log('chunk', chunk);

            raceCount += 1;
            totalRaceCount += 1;

            this.push(chunk)
            transformCallback()
          })
        )
    }

    return MultiStream(factory, { objectMode: true, highWaterMark: query.perPage })
  }
```

**Comment by stephenplusplus** [comment]
That's great to hear! And thanks for posting here, so others can see how mind-crippling streams are. At this point, I'm not even sure the people who wrote them fully understand them!

**Comment by techniq** [comment]
haha, no kidding.  They're like regex to me, as soon as I think I'm on top of them, I take a week off, and feel like I know half what I did last time I looked at them.

On a similar note, I have hopes the [observable](https://github.com/zenparsing/es-observable) / [web streams](https://jakearchibald.com/2016/streams-ftw/).  Haven't played with RxJS yet, but figure I will soon when I get into Angular 2.  I considered using it at one point (or Highland.js) but felt like it was overkill at the time for what I was doing and was trying to keep things as simple as possible.
